Test/tutorial/Popit.dfy

```diff
@@ -0,0 +1,963 @@
+/// that we are extracting with `var x :| x in popIt`
+
+  ghost function Pick<T>(move: multiset<T>): T
+    requires |move| != 0
+  {
+    var x :| x in move;
+    x
+  }
+
+/// Now, let's consider the moves. A move consists of considering a number of 'row' consecutive bubbles,
+/// and splitting it in two rows of index and row-amount-index rows
+
+  datatype Move =
+    Split(row: int, index: int, amount: int)
+
+/// Now, we want to have a say on whether the move is valid for a given PopIt
+/// and how to apply it if it is the case.
+/// For that, we define a datatype `PopItMove` that will be handy and contain a PopIt game and a move:
+
+  datatype PopItMove = PopItMove(popIt: PopIt, move: Move) {
+
+/// What does it mean for this move to be valid for the PopIt?
+/// it means that
+/// - that the elements we consider do exist in the original PopIt (`row in popIt`)
+/// - that the elements we are removing between index and index+amount are within the range
+
+    predicate Valid() {
+        match move {
+        case Split(row, index, amount) =>
+            row in popIt && 0 <= index < index + amount <= row
+        }
+    }
+
+/// Now that we defined this notion of move validity,
+/// we can define how applying this move results in a new PopIt.
+/// Note that, because we defined a datatype, we don't need to repeat the PopIt and move argument
+/// This makes signatures smaller and more manageable
+
+    function Apply(): PopIt
+      requires Valid()
+    {
+      match move {
+      case Split(row, index, amount) =>
+
+/// In every case, Dafny will verify that we return a valid PopIt, e.g. no element is zero.
+/// In the first case, when we remove all the elements of a row, we simply delete it
+
+        if amount == row then
+        popIt - multiset{row}
+
+/// In the second case, if we remove elements on either side,
+/// it means that we are replacing the row with one smaller row
+
+        else if index == 0 || index + amount == row then
+        popIt - multiset{row} + multiset{row-amount}
+
+/// In the third case, if we remove elements from the middle
+/// we create two smaller rows
+
+        else
+        popIt - multiset{row} + multiset{index} + multiset{row-index-amount}
+      }
+    }
+
+/// Now we can prove our termination metric. If a move can be applied to a PopIt,
+/// we show that its number of elements decreases
+/// That way, we can prove that all our recursive functions terminate.
+
+    lemma {:axiom} Decreases()
+        requires Valid()
+        ensures NumberOfElements(Apply()) < NumberOfElements(popIt)
+    {
+      match move { 
+      case Split(row, index, amount) =>
+        if amount == row {
+
+/// We use a Calc statement to compute how the number of bubbles evolves after application,
+/// depending on the type of move.
+/// "calc" takes expressions and it will try to prove that each expressions equals the next,
+/// unless another operator like "<" is provided.
```

Comment:
```suggestion
/// `calc` takes expressions and it will try to prove that each expressions equals the next,
/// unless another operator like `<` is provided.
```